src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+          ErrorVisitorT(WrappedAllowedErrorsT)>>...>;
+      // OK, now we know about all errors next continuation must take
+      // care about. If Visitor handled everything and the Value Func
+      // doesn't return any, we'll finish with errorator<>::future
+      // which is just vanilla seastar::future – that's it, next cont
+      // finally could use `.then()`!
+      using Futurator = \
+        typename ReturnErrorator::template futurize<ValueFuncResult>;
+      // `seastar::futurize`, used internally by `then_wrapped()`, wraps
+      // any non-`seastar::future` in `seastar::future`. We really don't
+      // want to end with e.g. `seastar::future<errorator::future<...>>`,
+      // so the lambda converts errorated future into plain one and here
+      // we're converting it back. As there is absolutely no difference
+      // between these types from the memory layout's POV (casting could
+      // be used instead), I count on compiler's ability to elide copies.
+      return typename Futurator::type{ this->then_wrapped(
+        [ valfunc = std::forward<ValueFuncT>(valfunc),
+          errfunc = std::forward<ErrorVisitorT>(errfunc)
+        ] (auto future) mutable {
+          if (__builtin_expect(future.failed(), false)) {
+            maybe_handle_error_t<ErrorVisitorT, Futurator> maybe_handle_error(
+              std::forward<ErrorVisitorT>(errfunc),
+              std::move(future).get_exception()
+            );
+            (maybe_handle_error.template operator()<WrappedAllowedErrorsT>() , ...);
+            return plainify(std::move(maybe_handle_error).get_result());
+          } else {
+            return plainify(Futurator::apply(std::forward<ValueFuncT>(valfunc),
+                                             std::move(future).get()));
+          }
+        })};
+    }
+
+    // taking ErrorFuncOne and ErrorFuncTwo separately from ErrorFuncTail
+    // to avoid SFINAE
+    template <class ValueFunc,
+              class ErrorFuncOne,
+              class ErrorFuncTwo,
+              class... ErrorFuncTail>
+    auto safe_then(ValueFunc&& value_func,
+                   ErrorFuncOne&& error_func_one,
+                   ErrorFuncTwo&& error_func_two,
+                   ErrorFuncTail&&... error_func_tail) {
+      return safe_then(
+        std::forward<ValueFunc>(value_func),
+        composer(std::forward<ErrorFuncOne>(error_func_one),
+                 std::forward<ErrorFuncTwo>(error_func_two),
+                 std::forward<ErrorFuncTail>(error_func_tail)...));
+    }
+
+    template <class Func>
+    void then(Func&&) = delete;
+
+
+  private:
+    // for the sake of `plainify()` let any errorator convert errorated
+    // future into plain one.
+    template <class... AnyWrappedAllowedErrorsT>
+    friend class errorator;
+
+    base_t&& as_plain_future() && {
+      return std::move(*this);
+    }
+  };
+
+  // the visitor that forwards handling of all errors to next continuation
+  struct pass_further {
```

Comment:
I think this behavior could be replaced with my safe_then/safe_then_exhaustive concept above.  In that case, safe_then with no error handlers would have the effect of the current safe_then with pass_further.